test/lit/ctor-eval/return_call.wast

```diff
@@ -0,0 +1,529 @@
+  (global $g1 (export "g1") (mut i32) (i32.const 0))
+
+  (func $test (export "test")
+    (global.set $g1
+      (i32.const 1)
+    )
+    (return_call $import)
+  )
+)
+
+;; CHECK:      (export "g1" (global $g1))
+
+;; CHECK:      (export "test" (func $test_2))
+
+;; CHECK:      (func $test_2 (type $0)
+;; CHECK-NEXT:  (return_call $import)
+;; CHECK-NEXT: )
+(module
+  ;; Return call to import with params
+
+  ;; CHECK:      (type $0 (func (param i32)))
+
+  ;; CHECK:      (type $1 (func))
+
+  ;; CHECK:      (import "env" "import" (func $import (type $0) (param i32)))
+  (import "env" "import" (func $import (param i32)))
+
+  ;; CHECK:      (global $g1 (mut i32) (i32.const 1))
+  (global $g1 (export "g1") (mut i32) (i32.const 0))
+
+  (func $test (export "test")
+    (global.set $g1
+      (i32.const 1)
+    )
+    (return_call $import
+      (i32.add
+        (i32.const 40)
+        (i32.const 2)
+      )
+    )
+  )
+)
+
+;; CHECK:      (export "g1" (global $g1))
+
+;; CHECK:      (export "test" (func $test_2))
+
+;; CHECK:      (func $test_2 (type $1)
+;; CHECK-NEXT:  (return_call $import
+;; CHECK-NEXT:   (i32.const 42)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+(module
+  ;; Chain of return calls ending in import
+
+  ;; CHECK:      (type $0 (func (param i32)))
+
+  ;; CHECK:      (type $1 (func))
+
+  ;; CHECK:      (import "env" "import" (func $import (type $0) (param i32)))
+  (import "env" "import" (func $import (param i32)))
+
+  ;; CHECK:      (global $g1 (mut i32) (i32.const 1))
+  (global $g1 (export "g1") (mut i32) (i32.const 0))
+
+  ;; CHECK:      (global $g2 (mut i32) (i32.const 2))
+  (global $g2 (export "g2") (mut i32) (i32.const 0))
+
+  (func $test (export "test")
+    (global.set $g1
+      (i32.const 1)
+    )
+    (return_call $test2
+      (i32.const 40)
+    )
+  )
+
+  (func $test2 (param i32)
+    (global.set $g2
+      (i32.const 2)
+    )
+    (return_call $import
+      (i32.add
+        (i32.const 2)
+        (local.get 0)
+      )
+    )
+  )
+)
+
+;; CHECK:      (export "g1" (global $g1))
+
+;; CHECK:      (export "g2" (global $g2))
+
+;; CHECK:      (export "test" (func $test_3))
+
+;; CHECK:      (func $test_3 (type $1)
+;; CHECK-NEXT:  (return_call $import
+;; CHECK-NEXT:   (i32.const 42)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+(module
+  ;; Return call to a function that can only be partially evaluated.
+  ;; CHECK:      (type $0 (func))
+
+  ;; CHECK:      (import "env" "import" (func $import (type $0)))
+  (import "env" "import" (func $import))
+
+  ;; CHECK:      (global $g1 (mut i32) (i32.const 1))
+  (global $g1 (export "g1") (mut i32) (i32.const 0))
+
+  ;; CHECK:      (global $g2 (mut i32) (i32.const 2))
+  (global $g2 (export "g2") (mut i32) (i32.const 0))
+
+  (func $test (export "test")
+    (global.set $g1
+      (i32.const 1)
+    )
+    (return_call $test2)
+  )
+
+  (func $test2
+    (global.set $g2
+      (i32.const 2)
+    )
+    (call $import)
```

Comment:
This seems to overlap wit the previous testcase, which also had partial eval of the second function. But this call was a return-call there.

Reply:
Right, in the previous test case the roll of the import is to be return-called so that it actually ends up as the current `func` at the end of evaluation. In contrast, the import in this test is just there to abort the eval and leave the current `func`, `$test2`, partially evaluated.